### PR TITLE
Makefile refactored (#16)

### LINKED ISSUE #16: fix: Refactor makefile
- [x] Separate directories and sub directories
- [x] Remove recursive deletions
- [x] Remove unwanted deletions
- [x] Add debug target
- [x] Refactor test execution and coverage display
- [x] Fix .PHONY
- [x] Add debug and test binaries to fclean
- [x] Make Makefile great again

### PR DESCRIPTION
**Pull request by RezaRahemtola**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -55,31 +78,28 @@ LDFLAGS		=	-L lib/my/
...
+$(TEST_BINARY): CFLAGS += -ftest-coverage -fprofile-arcs
+$(TEST_BINARY):	$(TEST_OBJ)
+			$(MAKE) -C lib/my
+			$(CC) -o $(TEST_BINARY) $(TEST_OBJ) $(LDFLAGS) $(LDLIBS)
 
 clean:
-			make clean -C lib/my/
 			$(RM) $(OBJ)
-			$(RM) *.gcda
-			$(RM) *.gcno
```
Comment (RezaRahemtola) [inline Makefile:67]: please clean `.gcno` and `.gcda` files, at least in an intern rule `clean_tests` or something like that

**Review by RezaRahemtola — COMMENTED** [review]
Currently completely instable:
`make tests_run` generate overwriting errors on gcda/gcno files + warnings on some tests

Maybe you can put OBJ files generated for tests in a dedicated folder to avoid conflicts between obj files with test flags and files without

**Comment by Croos3r** [comment]
> Currently completely instable: `make tests_run` generate overwrtiging errors on gcda/gcno files + warnings on some tests
> 
> Maybe you can put OBJ files generated for tests in a dedicated folder to avoid conflicts between obj files with test flags and files without

Or we could just  `make re` when we need to build a different way

**Comment by RezaRahemtola** [comment]
> Or we could just `make re` when we need to build a different way

Not really clear IMO, if I call `make tests_run` for example I expect it to launch the unit tests and do the needed stuff in the background for me, why bother adding `re` every time

**Comment by Croos3r** [comment]
> > Or we could just `make re` when we need to build a different way
> 
> Not really clear IMO, if I call `make tests_run` for example I expect it to launch the unit tests and do the needed stuff in the background for me, why bother adding `re` every time

`make tests_run` will not make any complaints, `tests_run` calls `fclean` exactly like `re` and `debug`

**Comment by Croos3r** [comment]
> > > Or we could just `make re` when we need to build a different way
> > 
> > 
> > Not really clear IMO, if I call `make tests_run` for example I expect it to launch the unit tests and do the needed stuff in the background for me, why bother adding `re` every time
> 
> `make tests_run` will not make any complaints, `tests_run` calls `fclean` exactly like `re` and `debug`

Embarassing, it, at the time, does not calling fclean

**Comment by RezaRahemtola** [comment]
> `make tests_run` will not make any complaints, `tests_run` calls `fclean` exactly like `re` and `debug`

Yes, but `make tests_run` followed by `make clean` will then cause both `make tests_run` or `make tests_func` fail for example, and IMO it should‘nt, tests rules must do what‘s needed to run the tests, which includes deleting obj/gcda/gcno files that could interfere

**Comment by Croos3r** [comment]
> > `make tests_run` will not make any complaints, `tests_run` calls `fclean` exactly like `re` and `debug`
> 
> Yes, but `make tests_run` followed by `make clean` will then cause both `make tests_run` or `make tests_func` fail for example, and IMO it should‘nt, tests rules must do what‘s needed to run the tests, which includes deleting obj/gcda/gcno files that could interfere

So I'll just remove gcda and gcno, tests_run currently rebuild and add re to dependencies of tests_func

**Comment by Croos3r** [comment]
Bam des mails

File: Makefile [hunk 2]
```diff
@@ -5,83 +5,123 @@
...
-fclean:		clean
-			make fclean -C lib/my/
-			$(RM) $(NAME)
-			$(RM) $(TEST)
+fclean:	clean
+			$(RM) $(BINARY) $(TEST_BINARY) $(DEBUG_BINARY)
 
-re:			fclean all
+clean_coverage:	fclean
+			@find \( -name '*.gcno' -o -name '*.gcda' \) -delete
```
Comment (RezaRahemtola) [inline Makefile]: could use `$(RM) *.gcno` and same for gcda files, bur otherwise LGTM

File: Makefile [hunk 3]
```diff
@@ -5,83 +5,123 @@
...
+
+$(TEST_BINARY): LDLIBS	+=	-lcriterion -lgcov
+$(TEST_BINARY): CFLAGS	+=	-ftest-coverage -fprofile-arcs
+$(TEST_BINARY):	$(TEST_OBJ)
+			$(MAKE) -C lib/my
+			$(CC) -o $(TEST_BINARY) $(TEST_OBJ) $(LDFLAGS) $(LDLIBS)
+
+$(DEBUG_BINARY):	CFLAGS	+=	-g
+$(DEBUG_BINARY):	$(OBJ)
+			$(MAKE) -C lib/my
```
Comment (RezaRahemtola) [inline Makefile]: and align instructions it's cleaner

**Comment by RezaRahemtola** [comment]
LGTM, great job !

**Comment by Croos3r** [comment]
> LGTM, great job !

stfu

**Review by Steci — APPROVED** [review]
Good Tea

**Review by EdenComp — APPROVED** [review]
LGTM tant que ça compile

**Pull request merged**